--- ./linux-5.15.0/mm/memory.c.orig	2025-04-09 22:24:28.178308847 -0600
+++ ./linux-5.15.0/mm/memory.c	2025-04-14 15:58:51.802097518 -0600
@@ -3875,6 +3875,7 @@
 	inc_mm_counter_fast(vma->vm_mm, MM_ANONPAGES);
 	page_add_new_anon_rmap(page, vma, vmf->address, false);
 	lru_cache_add_inactive_or_unevictable(page, vma);
+	handle_extents(vmf->address, page_to_pfn(page));
 setpte:
 	set_pte_at(vma->vm_mm, vmf->address, vmf->pte, entry);
 
