--- ./linux-5.15.0/mm/Makefile.orig	2025-04-09 22:29:02.898925941 -0600
+++ ./linux-5.15.0/mm/Makefile	2025-04-09 22:29:21.511239426 -0600
@@ -35,7 +35,7 @@
 CFLAGS_init-mm.o += $(call cc-disable-warning, initializer-overrides)
 
 mmu-y			:= nommu.o
-mmu-$(CONFIG_MMU)	:= highmem.o memory.o mincore.o \
+mmu-$(CONFIG_MMU)	:= highmem.o memory.o my_extent.o mincore.o \
 			   mlock.o mmap.o mmu_gather.o mprotect.o mremap.o \
 			   msync.o page_vma_mapped.o pagewalk.o \
 			   pgtable-generic.o rmap.o vmalloc.o
