--- /dev/null	2025-04-10 00:28:07.172000538 -0600
+++ ../project2/linux-5.15.0/include/linux/my_extent.h	2025-04-10 11:57:48.174298516 -0600
@@ -0,0 +1,38 @@
+#ifndef _LINUX_MY_EX_H
+#define _LINUX_MY_EX_H
+
+#include <linux/errno.h>
+
+#ifdef __KERNEL__
+
+#include <linux/mm.h>
+#include <linux/rbtree.h>
+#include <linux/list.h>
+
+struct list_extns {
+    struct list_head pg_list;
+    unsigned long pte;
+    long long vma;
+};
+
+struct rb_extns {
+    struct rb_node pg_tree;
+    struct list_head p_node;
+    size_t count;
+};
+
+extern pid_t do_extents_pid;
+
+void free_pg_list(struct list_head *head);
+void free_pg_tree(void);
+
+int compare_pg(struct rb_node **cursor, struct rb_extns *node, unsigned long long pte);
+struct rb_extns *pg_search(struct rb_root *root, unsigned long vma, unsigned long long pte);
+int pg_insert(struct rb_root *root, struct rb_extns *data);
+
+struct list_extns *create_extn(unsigned long vma, unsigned long long pte);
+void update_pg_extns(struct rb_extns *rb_node, unsigned long vma, unsigned long long pte);
+void add_extents(unsigned long vma, unsigned long long pte);
+
+#endif /* __KERNEL__ */
+#endif /* _LINUX_MY_EX_H */
