--- /dev/null	2025-04-10 19:48:58.816000561 -0600
+++ ../project2/linux-5.15.0/mm/my_extent.c	2025-04-10 20:31:29.536965193 -0600
@@ -0,0 +1,219 @@
+#include <linux/kernel.h>
+#include <linux/syscalls.h>
+#include <linux/my_extent.h>
+#include <linux/printk.h>
+
+pid_t do_extents_pid = -1;
+EXPORT_SYMBOL(do_extents_pid);
+
+// my rb root tree for page extns
+struct rb_root expg_root = RB_ROOT;
+size_t pt_count;
+
+#define LOG_MY_FUNCS
+
+#ifdef LOG_MY_FUNCS
+    #define is_empty(dummy, ...) ( sizeof( (char[]){#__VA_ARGS__} ) == 1 )
+    #define LOG_FUNCS(format, ...) do { \
+        printk(format "%s:%d", __func__, __LINE__); \
+        if (!is_empty(__VA_ARGS__)) printk(KERN_CONT, "::"); \
+        printk(KERN_CONT __VA_ARGS__); \
+        printk(KERN_CONT "\n"); \
+    } while(0)
+
+    #define LPRINTK(...) printk(__VA_ARGS__)
+#else
+    #define LOG_FUNCS(format, ...)
+    #define LPRINTK(...)
+#endif
+
+void free_pg_list(struct list_head *head) {
+    struct list_extns *curr = NULL;
+    struct list_extns *node = NULL;
+
+    LOG_FUNCS(KERN_INFO);
+    list_for_each_entry_safe(curr, node, head, pg_list) {
+        LPRINTK(KERN_CONT "%lu:%llu", curr->vma, curr->pte);
+        list_del(&curr->pg_list);
+        kfree(curr);
+
+        if (node) LPRINTK(KERN_CONT ",");
+    }
+    LPRINTK(KERN_CONT "\n");
+}
+
+void free_pg_tree(void) {
+    struct rb_node *root;
+    int iter = 0;
+
+    LOG_FUNCS(KERN_INFO);
+    while ((root = rb_first(&expg_root)) != NULL) {
+        struct rb_extns *node = rb_entry_safe(root, struct rb_extns, pg_tree);
+        if (node) {
+            LOG_FUNCS(KERN_INFO, "%d:%zu::", iter, node->count);
+            free_pg_list(&node->p_node);
+            kfree(node);
+        } else {
+            LOG_FUNCS(KERN_CRIT, "%d", iter);
+        }
+        rb_erase(root, &expg_root);
+        iter++;
+    }
+}
+
+int compare_pg(struct rb_node *cursor, struct rb_extns *node, unsigned long vma) {
+    struct list_extns *head = list_first_entry_or_null(&node->p_node, struct list_extns, pg_list);
+    struct list_extns *tail = list_last_entry(&node->p_node, struct list_extns, pg_list);
+
+    LOG_FUNCS(KERN_INFO, "%lu", vma);
+    if (head == NULL) {
+        LOG_FUNCS(KERN_CRIT, "Invalid entry while adding vma %lu\n", vma);
+        return -1; // invalid entry!!
+    }
+
+    if (head->vma > vma)
+        return 1; // continue searching left
+    else if (tail->vma < vma)
+        return 2; // continue searching right
+    else
+        return 0;
+}
+
+struct rb_extns *pg_search(struct rb_root *root, unsigned long vma) {
+    struct rb_node *cursor = root->rb_node;
+    int result;
+
+    LOG_FUNCS(KERN_INFO, "%lu", vma);
+    while(cursor) {
+        struct rb_extns *node = container_of(cursor, struct rb_extns, pg_tree);
+
+        LOG_FUNCS(KERN_INFO, "%lu", vma);
+        result = compare_pg(cursor, node, vma);
+
+        if (result == -1)
+            return NULL;
+        else if (result == 1)
+            cursor = cursor->rb_left;
+        else if (result == 2)
+            cursor = cursor->rb_right;
+        else
+            return node;
+    }
+    return NULL;
+}
+
+int pg_insert(struct rb_root *root, struct rb_extns *data) {
+    struct list_extns *head = list_first_entry_or_null(&data->p_node, struct list_extns, pg_list);
+    struct rb_node **cursor = &(root->rb_node);
+    struct rb_node *parent = NULL;
+    int result;
+
+    LOG_FUNCS(KERN_INFO);
+    if (head == NULL) return 0;
+
+    LOG_FUNCS(KERN_INFO, "%lu:%llu", head->vma, head->pte);
+    while(*cursor) {
+        struct rb_extns *node = container_of(*cursor, struct rb_extns, pg_tree);
+        parent = *cursor;
+
+        LOG_FUNCS(KERN_INFO, "%lu:%llu", head->vma, head->pte);
+        result = compare_pg(*cursor, node, head->vma);
+        if (result == 1)
+            cursor = &((*cursor)->rb_left);
+        else if (result == 2)
+            cursor = &((*cursor)->rb_right);
+        else
+            return 0;
+    }
+
+    rb_link_node(&data->pg_tree, parent, cursor);
+    rb_insert_color(&data->pg_tree, root);
+    LOG_FUNCS(KERN_INFO, "%lu:%llu", head->vma, head->pte);
+    return 1;
+}
+
+struct list_extns *create_extn(unsigned long vma, unsigned long long pte) {
+    struct list_extns *list_node = kmalloc(sizeof(struct list_extns), GFP_KERNEL);
+    LOG_FUNCS(KERN_INFO, "%lu:%llu", vma, pte);
+    list_node->vma = vma;
+    list_node->pte = pte;
+    return list_node;
+}
+
+void update_pg_extns(struct rb_extns *rb_node, unsigned long vma, unsigned long long pte) {
+    struct list_head *head = &rb_node->p_node;
+    struct list_extns *node = list_first_entry(head, struct list_extns, pg_list);
+    struct list_head *pos = NULL;
+
+    LOG_FUNCS(KERN_INFO, "%lu:%llu", vma, pte);
+    if (node->vma > vma) {
+        list_add_tail(&(create_extn(vma, pte))->pg_list, head);
+        rb_node->count++;
+        return;
+    }
+
+    node = list_last_entry(head, struct list_extns, pg_list);
+    if (node->vma < vma) {
+        list_add(&(create_extn(vma, pte))->pg_list, head->prev);
+        rb_node->count++;
+        return;
+    }
+
+    list_for_each(pos, head) {
+        node = list_entry(pos, struct list_extns, pg_list);
+        if (node->vma != vma) continue;
+        node->vma = vma;
+        return;
+    }
+}
+
+void add_extents(unsigned long vma, unsigned long long pte) {
+    struct rb_extns *rb_node = kmalloc(sizeof(struct rb_extns), GFP_KERNEL);
+
+    LOG_FUNCS(KERN_INFO, "%lu:%llu", vma, pte);
+    rb_node->count = 1;
+    INIT_LIST_HEAD(&rb_node->p_node);
+    list_add(&create_extn(vma, pte)->pg_list, &rb_node->p_node);
+    pg_insert(&expg_root, rb_node);
+    pt_count++;
+}
+
+void handle_extents(unsigned long vma, unsigned long long pte) {
+    struct rb_extns *node = NULL;
+    pid_t pid = task_pid_nr(current);
+
+    if (pid ==  do_extents_pid) {
+        LOG_FUNCS(KERN_INFO, "%lu:%llu", vma, pte);
+
+        // currently only updates existing nodes, need to handle the other part too
+        node = pg_search(&expg_root, vma);
+        if (node)
+            update_pg_extns(node, vma, pte);
+        else
+            add_extents(vma, pte);
+    }
+}
+EXPORT_SYMBOL(handle_extents);
+
+SYSCALL_DEFINE0(enable_extents)
+{
+    pid_t pid = task_pid_nr(current);
+
+    LOG_FUNCS(KERN_INFO, "%d", pid);
+    do_extents_pid = pid;
+    pt_count = 0;
+    return 0;
+}
+
+SYSCALL_DEFINE0(disable_extents)
+{
+    pid_t pid = task_pid_nr(current);
+    LOG_FUNCS(KERN_INFO, "%d had %zu pages", pid, pt_count);
+
+    if (pid == do_extents_pid) {
+        do_extents_pid = -1;
+        pt_count = 0;
+        free_pg_tree();
+    }
+    return 0;
+}
