--- /dev/null	2025-04-09 22:13:39.712000557 -0600
+++ ../project2/linux-5.15.0/include/linux/my_extent.h	2025-04-09 22:38:57.237006700 -0600
@@ -0,0 +1,25 @@
+#ifndef _LINUX_MY_EX_H
+#define _LINUX_MY_EX_H
+
+#include <linux/errno.h>
+
+#ifdef __KERNEL__
+#include <linux/mm.h>
+#include <linux/rbtree.h>
+#include <linux/list.h>
+
+struct pg_extns {
+    struct list_head list;
+    unsigned long pte;
+    unsigned long vma;
+};
+
+struct rb_extns {
+    struct rb_node r_node;
+    struct pg_extns p_node;
+    size_t count;
+};
+
+extern pid_t do_extents_pid;
+#endif /* __KERNEL__ */
+#endif /* _LINUX_MY_EX_H */
