--- /dev/null	2025-04-09 17:26:49.608000450 -0600
+++ ../project2/linux-5.15.0/mm/my_extent.c	2025-04-09 19:51:19.052055593 -0600
@@ -0,0 +1,34 @@
+#include <linux/kernel.h>
+#include <linux/my_extent.h>
+#include <linux/printk.h>
+#include <linux/syscalls.h>
+
+pid_t do_extents_pid = -1;
+EXPORT_SYMBOL(do_extents_pid);
+
+// my rb root tree for page extns
+struct rb_root pg_tree = RB_ROOT;
+
+SYSCALL_DEFINE0(enable_extents)
+{
+    pid_t pid = task_pid_nr(current);
+    do_extents_pid = pid;
+    return 0;
+}
+
+SYSCALL_DEFINE0(disable_extents)
+{
+    pid_t pid = task_pid_nr(current);
+    if (pid == do_extents_pid) {
+        do_extents_pid = -1;
+    }
+    return 0;
+}
+
+void handle_extents(unsigned long vma, long long pte) {
+    pid_t pid = task_pid_nr(current);
+    if (pid ==  do_extents_pid) {
+        printk(KERN_INFO "%d::vma %lu assigned pte %08llx\n", do_extents_pid, vma, pte);
+    }
+}
+EXPORT_SYMBOL(handle_extents);
