--- ./linux-5.15.0/arch/x86/entry/syscalls/syscall_64.tbl.orig	2025-04-09 22:21:52.415695977 -0600
+++ ./linux-5.15.0/arch/x86/entry/syscalls/syscall_64.tbl	2025-04-12 18:56:32.304616741 -0600
@@ -343,6 +343,10 @@
 332	common	statx			sys_statx
 333	common	io_pgetevents		sys_io_pgetevents
 334	common	rseq			sys_rseq
+335 common  enable_extents  sys_enable_extents
+336 common  disable_extents sys_disable_extents
+337 common  get_ex_count    sys_get_ex_count
+338 common  print_ex_pid   sys_print_ex_pid
 # don't use numbers 387 through 423, add new calls after the last
 # 'common' entry
 424	common	pidfd_send_signal	sys_pidfd_send_signal
