--- ../project2/linux-5.15.0/arch/x86/entry/syscalls/syscall_64.tbl.orig	2025-04-09 14:52:13.237978306 -0600
+++ ../project2/linux-5.15.0/arch/x86/entry/syscalls/syscall_64.tbl	2025-04-09 14:52:13.277978934 -0600
@@ -343,6 +343,8 @@
 332	common	statx			sys_statx
 333	common	io_pgetevents		sys_io_pgetevents
 334	common	rseq			sys_rseq
+335 common  enable_extents  sys_enable_extents
+336 common  disable_extents  sys_disable_extents
 # don't use numbers 387 through 423, add new calls after the last
 # 'common' entry
 424	common	pidfd_send_signal	sys_pidfd_send_signal
