--- ../project2/linux-5.15.0/include/linux/syscalls.h.orig	2025-04-09 14:52:13.241978369 -0600
+++ ../project2/linux-5.15.0/include/linux/syscalls.h	2025-04-09 15:06:04.787059346 -0600
@@ -1279,6 +1279,9 @@
  */
 asmlinkage long sys_ni_syscall(void);
 
+/* mm/my_extent.c */
+asmlinkage long sys_enable_extents();
+asmlinkage long sys_disable_extents();
 #endif /* CONFIG_ARCH_HAS_SYSCALL_WRAPPER */
 
 
